### DISCUSSION TITLE
PicoGK using python

### DISCUSSION THREAD

**Discussion by jigarmech** [leap71/PicoGK#26] [Q&A]
Hello Team,

I am using pythonnet module for the first time. I have tried to run the first example provided here on PicoGK. It did run without any error using python but the viewer that should show the Boolean cases isn't showing. Any idea where I could have gone wrong?

Input
```

from pythonnet import load
load("coreclr")
import clr
clr.AddReference('PicoGK')
import PicoGK

try:
    PicoGK.Library.Go(0.5, PicoGK.BooleanShowCase.Task())
except Exception as e:
    print(e)

```
Output
```
BooleanShowCase
Press any key to continue . . .
```

**Comment by LinKayser** [comment]
I can only speculate that maybe, somehow the window cannot be created, so it's not showing. You could check if the PicoGK logfile is actually created and written in your Documents folder. This could hint at what is going on.

Best,
Lin

**Reply by jigarmech** [reply]
Thank you for the assistance, Lin. The log file was not in the Documents folder, I will investigate further to understand the issue. Additionally, I plan to delve into C# exploration. 
Your contribution has been immensely valuable.

Best Regards,
Jigar

**Comment by yvanblanchard** [comment]
Any update on this issue @jigarmech ?

**Reply by jigarmech** [reply]
No, I didn't look further into it. Started working with C# instead. If you come across a solution, do share!

**Reply by virtuamechanica** [reply]
Hello @jigarmech 
I plan to work on python bindings for picoGk. Could you please share your work with pythonNet , since i am not so familiar with c# bindings , avoiding starting from scratch.
Thank you for your help!

**Reply by KofiPhD2019** [reply]
Any update on how it went with python and PicoGK? I am planning on using python too. But if it does not work out, I will learn C#.

**Comment by ghedo44** [comment]
Hi all, I built a community Python package for PicoGK runtime called pycogk.

PyPI: [https://pypi.org/project/pycogk](https://pypi.org/project/pycogk)
Source: [https://github.com/ghedo44/PycoGK](https://github.com/ghedo44/PycoGK)

It provides Python bindings and a viewer workflow.
If anyone from this thread wants to try it and report issues (especially around viewer/runtime setup), I’d really appreciate feedback.

Note: this is community-maintained and not an official PicoGK release.

**Reply by yvanblanchard** [reply]
Thanks @ghedo44  !!

**Reply by KofiPhD2019** [reply]
Any feedback from anyone? I am considering using Python too.

**Reply by LinKayser** [reply]
I (obviously) have a comment. While I appreciate the Python effort, we made a very conscious decision not to use Python, despite its wide user base. Python is nice for scripting, but not great for Computational Engineering. My book project, Coding for Engineers, discusses this in the first chapter. https://picogk.org/coding-for-engineers/

**Comment by joshmelnykcroft** [comment]
As a Python user, I learned C# just for picoGK. Try it! It's not as hard a
transition as you might think. The ecosystem around C# has been improving
as well.

**Reply by KofiPhD2019** [reply]
I have downloaded the lastest version of picoGK but I do not know what to do next. The documentation has been removed. I cannot follow the tutorial on Lin's documentation page.

**Comment by Borderliner** [comment]
@joshmelnykcroft Check this out: [PicoPie](https://github.com/Borderliner/PicoPie)

**Comment by LinKayser** [comment]
Dear @Borderliner — I commend the effort to bring PicoGK-style geometry work to Python. This looks like a high-quality and substantial project, and it will likely help more people experiment with this type of geometry.

My reservations about Python for serious Computational Engineering work remain: performance matters, but more importantly, strong typing and explicit software architecture become paramount in non-trivial engineering projects. That said, I recognize the value of making this accessible to the Python ecosystem.

I do have one request: please keep the naming clearly separated. Since this is not an official LEAP 71 project and is not supported by us, we would ask that you consistently use `PicoPie` rather than `PicoGK` in package names, namespaces, and type names where possible, to avoid confusion.

Please also note that `PicoGK.Shapes` is a namespace inside PicoGK and should not be introduced as a stand-in for ShapeKernel.

Keep up the good work.

Lin

**Reply by Borderliner** [reply]
@LinKayser Thank you for your review.
To address some of your concerns, we provide native loops (fastloop) which significantly improves looping and falls back to pure-python loops if not available (this is the most important place that Python would hit a performance wall compared to C#). All the bindings are directly to the C ABI so the performance does not suffer a noticeable hit. We also provide Pythonic type hints on our bindings so the API is clear from the user's side if they have type linters set up (full coverage).

But the namespacing is a real issue here. That was done to make it easier for Python users to hop onto PicoGK, but clearly I didn't think it through enough that it might cause issues with upstream, sorry about that. This is a high priority task and I'll jump on to it.

Thanks for making such an amazing software. I'd be happy to help maintaining official Python bindings if that's your direction. Honestly Python is very approachable to a broad audience other than software engineers, and I let PicoPie be a testimony of its performance and feature parity.

**Reply by Borderliner** [reply]
Just to show how the linting and type hinting looks like (e.g. on Emacs):
<img width="1271" height="715" alt="Screenshot from 2026-07-02 00-59-10" src="https://github.com/user-attachments/assets/38f29880-d1a5-4946-be89-cd7ec64ab609" />

**Reply by LinKayser** [reply]
Sure — but strong typing is something different altogether. 

When using physics, for example, it's important that, say 'Pressure' is incompatible with 'Acceleration'. We even go so far as defining 'ChamberPressure', so that you don't accidentally use it for an input that requires a different kind of pressure.

Python is great for simple scripts — but there is a reason why most large software projects are not written in Python. Its widespread use in Machine Learning is a bit accidental (and all complex/performance-critical stuff is not in Python).  

By the way LLM agents benefit greatly from feedback from strong typing. 'ChamberPressure' is more expressive than 'float'.

You can work around many of these things in Python, but C# enforces this way of thinking, which is a good thing. 

I should probably write a book chapter about this.

Lastly, don't underestimate how much heavy loading ends up outside the PicoGK runtime. Multithreading is important and JIT performance is much better on .NET.

But, again, great work on the port. 

I'd say though, instead of putting all this effort, why not give C# a go? 

It's a very elegant language, and it was chosen for a good reason (I looked at a lot of options before making the decision).

Best,
Lin

**Reply by Yvan-iComat** [reply]
@LinKayser 
I fully agree.
The only drawback with .NET (compared to Python) I see here is the lack of libraries. Most of them are in C++ , Python , or Rust. Those three are very easily compatible each other (meaning with a thin wrapper only).

**Reply by LinKayser** [reply]
Yes, indeed. Some libraries are not available of .NET. We sometimes execute an external Python process to run certain queries and then feed back into the CEM. Would be lovely to avoid this. I think with the recent advances in agentic coding, a lot of this stuff will end up on .NET as well, as it gets easier to port.

Lin

**Reply by Borderliner** [reply]
@LinKayser Unfortunately not many scientists outside of computer science are flexible enough to learn a new programming language. We didn't make Python simply because it was popular out there, but because our team of scientists only knew Python and transitioning them over to C# which is strongly-typed would be very time consuming and inefficient. So even if C# is technically a superior choice, Python offers the least friction for many teams. We built the Python port because we desperately needed it, we didn't do it just for the love of the game.

Thank you for your words of encouragement.
